docs/configure.rst
.. Hey, Emacs this is -*- rst -*-

   This file follows reStructuredText markup syntax; see
   http://docutils.sf.net/rst.html for more information.

.. configuration:

.. _CORS: https://en.wikipedia.org/wiki/Cross-origin_resource_sharing
.. _Flask: http://flask.pocoo.org/
.. _JWT: https://python-jose.readthedocs.io/en/latest/index.html
.. _engines: https://docs.sqlalchemy.org/en/latest/core/engines.html

=================
  Configuration
=================

All the information regarding the system is stored in a configuration file. The default
configuration file is stored in your home directory:
``~/.acpy/api.config`` but you can specify a different location from the command line
with the `-c` option.

If no configuration file is found, it will copy an example configuration file in ``~/.acpy/api.config``.
The example contains placeholders for all required values.

General settings
=================
The following parameters are available as general settings:

1. CORS_ can be enabled, by default it is disabled
2. ``secret`` for Flask_, autogenerated when the config file is created
3. ``port`` for the api, default is 8080
4. ``run_time`` stores the runtime information in a file, default is ``~/.acpy/run_time.data``

Logging settings
=================
The following parameters are available for logging:

1. ``log_file`` specifies the location of the logfile, default is ``~/.acpy/acpy.log``. If left empty, no log file is generated.
2. ``max_bytes`` is the maximum size of a logfile before it is rotated, default is 2.5MB.
3. ``backup_count`` is the maximum amount of logfiles to keep of max_bytes, default is 5.

The admin account
==================
The default admin account behaves as a service account:

1. ``access`` is the access code for the admin account (analog to username)
2. ``secret`` is the secret code for the admin account (analog to password) the password needs to be passed as sha256.

JWT token settings
===================
Users and services request a token by calling the ``login`` service. This service returns a JWT_ token.

1. ``issuer`` is the name of the token issuer, this should reflect the URI of the API.
2. ``secret`` is the secret that is used for encoding the token, if left empty the FLASK_ shared secret is used.
3. ``lifetime`` is the token lifetime in seconds, default is 3600.
4. ``algorithm`` is the encryption algorithm for the token, default is HS256.


Database settings
==================
All data is stored in a database, the connection needs to be specified in the configuration. By default we store to memory.

1. ``connection`` specifies the database connection, default is sqlite://

* example of a mariadb connection: ``mysql://scott:tiger@localhost/test``
* example of a postgres connection: ``postgresql://scott:tiger@localhost/mydatabase``

We use SQLAlchemy, so for all options see engines_

LDAP settings
==============
For authenticating users configure an LDAP compliant connection.

1. ``host`` ldap server without protocol.
2. ``port`` ldap port
3. ``ssl`` secure connection
4. ``base_dn`` root for the users
5. ``rdn_attr`` relative distinguished name (usually ``uid`` or ``cn``)
6. ``login_attr`` what field to use for username.
7. ``bind_user`` user account that has read access on the ``base_dn``
8. ``bind_pass`` password for ``bind_user``


config.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-#
#
#
# Copyright (C) 2018 University of Zurich. All rights reserved.
#
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 3 of the License, or (at your
# option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA

import configparser
import logging
import random

from pathlib import Path
from os.path import exists, dirname, expandvars, expanduser

allowed_chars = u'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'


class DefaultConfig(object):

    @staticmethod
    def create():
        config = configparser.ConfigParser()
        config.add_section('general')
        config.set('general', 'CORS', 'False')
        config.set('general', 'secret', ''.join(random.choice(allowed_chars) for c in range(14)))
        config.set('general', 'port', '8080')
        config.set('general', 'run_time', expandvars(expanduser('~/.acpy/run_time.data')))

        config.add_section('logging')
        config.set('logging', 'log_file', expandvars(expanduser('~/.acpy/acpy.log')))
        config.set('logging', 'max_bytes', '2621440')
        config.set('logging', 'backup_count', '5')

        config.add_section('admin')
        config.set('admin', 'access', ''.join(random.choice(allowed_chars) for c in range(12)))
        config.set('admin', 'secret', ''.join(random.choice(allowed_chars) for c in range(24)))

        config.add_section('token')
        config.set('token', 'issuer', 'com.example.accounting')
        config.set('token', 'secret', ''.join(random.choice(allowed_chars) for c in range(14)))
        config.set('token', 'lifetime', '3600')
        config.set('token', 'algorithm', 'HS256')

        config.add_section('database')
        config.set('database', 'connection', 'sqlite://')

        config.add_section('authentication')
        config.set('authentication', 'host', 'localhost')
        config.set('authentication', 'port', '636')
        config.set('authentication', 'ssl', 'True')
        config.set('authentication', 'base_dn', 'OU=Domain Users,DC=example,DC=com')
        config.set('authentication', 'rdn_attr', 'uid')
        config.set('authentication', 'login_attr', 'mail')
        config.set('authentication', 'bind_user', 'ldap_admin')
        config.set('authentication', 'bind_pass', 'ldap_admin_pass')

        return config


class Config(object):

    logger = logging.getLogger(__name__)

    def __init__(self, config_file=None, create=True):
        self.config = DefaultConfig().create()
        if create:
            self.write(expandvars(expanduser(config_file)))
        if config_file and exists(expandvars(expanduser(config_file))):
            self.config.read(expandvars(expanduser(config_file)))

    def write(self, path):
        if not exists(path):
            self.logger.debug("writing new config to {0}".format(path))
            if not exists(dirname(path)):
                Path.mkdir(Path(dirname(path)), parents=True)
            with open(path, 'w') as cf:
                self.config.write(cf)
        else:
            self.logger.debug("already found config at {0}, using it".format(path))

    def _fetch(self, section):
        result = {}
        options = self.config.options(section)
        for option in options:
            try:
                result[option] = self.config.get(section, option)
            except configparser.Error as e:
                self.logger.error("exception reading config on {0} ({1})".format(option, e))
                result[option] = None
        return result

    def general(self):
        return self._fetch('general')

    def logging(self):
        return self._fetch('logging')

    def admin(self):
        return self._fetch('admin')

    def token(self):
        return self._fetch('token')

    def database(self):
        return self._fetch('database')

    def authentication(self):
        return self._fetch('authentication')

    def update(self, section, option, value):
        self.config.set(section, option, value)
